### PR TITLE
[BugFix] Fix ReferenceGenerator Unions and Choices

### PR DESCRIPTION
**Pull request by deeleeramone**
1. **Why**?:

    - `reference.json` is creating too many Unions
    - "choices" for a provider contaminate other providers.
    - Expose "period" and "interval" parameters to `reference.json` and `openapi.json`

2. **What**?:

Before:
before1

After:

after1

Before:

before2

After:

Screenshot 2024-07-22 at 7 07 05 PM

3. **Impact**:

    - Where 'multiple_items_allowed', and there are defined choices, they are moved to `__json_schema_extra__` instead of defining under `Field`.
    - Less redundant type references.
    - "choices" for one provider do not force those "choices" on other providers.
    - `reference.json` more closely resembles `openapi.json`.
    - Financial statement "period" choices are now in `openapi.json`.
    - `x.price.historical` "interval" choices are now in `openapi.json`.

4. **Testing Done**:

    - Ran PackageBuilder tests.
    - Manual inspection of docstrings and `reference.json`
    - Check CLI `--help` dialogues.

The schema follows: Param -> schema -> (title has names of all providers) provider (if there are defined choices) -> choices

Screenshot 2024-07-24 at 2 27 23 PM

### CODE REVIEW

**Comment by IgorWounds** [comment]
@andrewkenreich can you take a look if this breaks anything on Terminal Pro?

File: openbb_platform/core/openbb_core/app/provider_interface.py [hunk 1]
```diff
@@ -253,14 +253,22 @@ def _create_field(
         annotation = field.annotation
 
         additional_description = ""
+        choices: Dict = {}
```
Comment (hjoaquim) [inline openbb_platform/core/openbb_core/app/provider_interface.py:256]: this variable name is misleading as it doesn't take only the choices but the extra info (being it choices and multiple items allowed info)

File: openbb_platform/core/openbb_core/app/provider_interface.py [hunk 2]
```diff
@@ -253,14 +253,22 @@ def _create_field(
...
         if extra := field.json_schema_extra:
-            providers = []
-            for p, v in extra.items():  # type: ignore[union-attr]
+            providers: List = []
+            for p, v in extra.items():  # type: ignore
                 if isinstance(v, dict) and v.get("multiple_items_allowed"):
                     providers.append(p)
+                    choices[p] = {"multiple_items_allowed": True, "choices": v.get("choices")}  # type: ignore
```
Comment (hjoaquim) [inline openbb_platform/core/openbb_core/app/provider_interface.py:262]: why are we assuming the `multiple_items_allowed` will be true here? shouldn't it be getting the value as it was before? `v.get("multiple_items_allowed")`

File: openbb_platform/core/openbb_core/app/provider_interface.py [hunk 3]
```diff
@@ -253,14 +253,22 @@ def _create_field(
...
                 elif isinstance(v, list) and "multiple_items_allowed" in v:
                     # For backwards compatibility, before this was a list
                     providers.append(p)
+                    choices[p] = {"multiple_items_allowed": True, "choices": None}  # type: ignore
+                elif isinstance(v, dict) and v.get("choices"):
+                    choices[p] = {
+                        "multiple_items_allowed": False,
+                        "choices": v.get("choices"),
```
Comment (hjoaquim) [inline openbb_platform/core/openbb_core/app/provider_interface.py:270]: does it make sense to have the choices here? are they needed for the command's execution?

File: openbb_platform/core/openbb_core/app/static/package_builder.py [hunk 4]
```diff
@@ -1742,25 +1754,48 @@ def get_paths(cls, route_map: Dict[str, BaseRoute]) -> Dict[str, Dict[str, Any]]
...
                             standard_model, "QueryParams", provider
                         )
                     )
+
                     # Adds provider specific data fields to the reference
                     reference[path]["data"][provider] = cls._get_provider_field_params(
                         standard_model, "Data", provider
                     )
+
+                    # Remove choices from 'standard' if choices for a parameter exist
```
Comment (hjoaquim) [inline openbb_platform/core/openbb_core/app/static/package_builder.py:1776]: if there's standard choices why do we remove it in favor of provider choices? i'd say standard should prevail

File: openbb_platform/core/openbb_core/provider/standard_models/spot.py [hunk 5]
```diff
@@ -32,7 +32,6 @@ class SpotRateQueryParams(QueryParams):
     category: str = Field(
         default="spot_rate",
         description="Rate category. Options: spot_rate, par_yield.",
-        json_schema_extra={"choices": ["par_yield", "spot_rate"]},
```
Comment (hjoaquim) [inline openbb_platform/core/openbb_core/provider/standard_models/spot.py:35]: why removing the standard choices?

File: openbb_platform/providers/alpha_vantage/openbb_alpha_vantage/models/equity_historical.py [hunk 6]
```diff
@@ -28,7 +28,10 @@ class AVEquityHistoricalQueryParams(EquityHistoricalQueryParams):
     Source: https://www.alphavantage.co/documentation/#time-series-data
     """
 
-    __json_schema_extra__ = {"symbol": {"multiple_items_allowed": True}}
+    __json_schema_extra__ = {
```
Comment (hjoaquim) [inline openbb_platform/providers/alpha_vantage/openbb_alpha_vantage/models/equity_historical.py:31]: could this be done at the `Field` level instead of model? the `OpenBBField` is expecting choices to be there - this would then influence the CLI

image

**Review by hjoaquim — COMMENTED** [review]
Left couple of comments for clarification.

Altough the `choices` are now only available on the reference since they're defined at the model level instead of the field, it looks like the CLI is still able to pick them up since its defined on `reference.json`.

CLI behavior seems good and choices are not conflicting between providers:

image

The help dialog also seems good. Take into consideration that since `argparse` doesn't allow repeated arguments, the help dialog for country will have all the choices regardless the provider - the choices should not be validated at the CLI (by argparse) but instead at the Platform (by pydantic); which seems to be happening as the image above shows.

File: openbb_platform/core/openbb_core/provider/standard_models/equity_historical.py [hunk 7]
```diff
@@ -20,10 +20,6 @@ class EquityHistoricalQueryParams(QueryParams):
     """Equity Historical Price Query."""
 
     symbol: str = Field(description=QUERY_DESCRIPTIONS.get("symbol", ""))
-    interval: Optional[str] = Field(
```
Comment (hjoaquim) [inline openbb_platform/core/openbb_core/provider/standard_models/equity_historical.py:23]: This changes on the standard models, namely adding/removing arguments should be done in a different PR unless there's a reason to do it on this one - it seems unrelated.

Comment (deeleeramone) [reply openbb_platform/core/openbb_core/provider/standard_models/equity_historical.py:23]: This is all related because now these choices are actually in `openapi.json` and `reference.json`, where they were not previously.

File: openbb_platform/providers/alpha_vantage/openbb_alpha_vantage/models/equity_historical.py (hunk 6 above)
Comment (deeleeramone) [reply openbb_platform/providers/alpha_vantage/openbb_alpha_vantage/models/equity_historical.py:31]: No, for things to show up in `openapi.json` it must be defined in `__json_schema_extra__`. Not sure why, but that is the way it appears to work.

File: openbb_platform/core/openbb_core/provider/standard_models/spot.py (hunk 5 above)
Comment (deeleeramone) [reply openbb_platform/core/openbb_core/provider/standard_models/spot.py:35]: Because these shouldn't be in the standard model.

File: openbb_platform/core/openbb_core/app/static/package_builder.py (hunk 4 above)
Comment (deeleeramone) [reply openbb_platform/core/openbb_core/app/static/package_builder.py:1776]: Because this is a main crux of the issue, combining all choices into one list where there is no distinction between providers. It should always be grouped by provider where multiple  providers exist. "Standard choices" != "valid choices".

File: openbb_platform/core/openbb_core/app/provider_interface.py (hunk 2 above)
Comment (deeleeramone) [reply openbb_platform/core/openbb_core/app/provider_interface.py:262]: It's not an assumption, the `if` statement evaluates `v["multiple_items_allowed"]` for truthiness.

**Comment by deeleeramone** [comment]
> Left couple of comments for clarification.
> 
> Altough the `choices` are now only available on the reference since they're defined at the model level instead of the field, it looks like the CLI is still able to pick them up since its defined on `reference.json`.
> 
> CLI behavior seems good and choices are not conflicting between providers:

The behaviour in CLI that is not correct is how when you enter a country for EconDB that is not in the choices for OECD. Choices for OECD are not all valid choices for EconDB, and you can also enter countries far more flexibly in the EconDB provider - i.e, "us", "uk", "USA", "united_states".

**Comment by deeleeramone** [comment]
> Thelp dialog for country will have all the choices regardless the provider - the choices should not be validated at the CLI (by argparse) but instead at the Platform (by pydantic); which seems to be happening as the image above shows.

This is actually not what happens with CLI, otherwise you would be able to enter: `economy --country us --provider econdb`.

File: openbb_platform/core/openbb_core/app/provider_interface.py (hunk 1 above)
Comment (deeleeramone) [reply openbb_platform/core/openbb_core/app/provider_interface.py:256]: This is a local variable that for iterating over choices for each the provider.  The only thing ever assigned to it are choices.

**Comment by andrewkenreich** [comment]
> @andrewkenreich can you take a look if this breaks anything on Terminal Pro?

I think we are good - but we can take care of it when we merge this too - wont affect us right away

File: openbb_platform/core/openbb_core/app/provider_interface.py (hunk 1 above)
Comment (hjoaquim) [reply openbb_platform/core/openbb_core/app/provider_interface.py:256]: is it? line 262 is `choices[p] = {"multiple_items_allowed": True, "choices": v.get("choices")}` 
sry if i'm misinterpreting

File: openbb_platform/providers/alpha_vantage/openbb_alpha_vantage/models/equity_historical.py (hunk 6 above)
Comment (hjoaquim) [reply openbb_platform/providers/alpha_vantage/openbb_alpha_vantage/models/equity_historical.py:31]: apparently is fine since it's allso available on the reference.json

File: openbb_platform/core/openbb_core/app/static/package_builder.py (hunk 4 above)
Comment (hjoaquim) [reply openbb_platform/core/openbb_core/app/static/package_builder.py:1776]: yap, gotcha

File: openbb_platform/core/openbb_core/provider/standard_models/equity_historical.py (hunk 7 above)
Comment (hjoaquim) [reply openbb_platform/core/openbb_core/provider/standard_models/equity_historical.py:23]: fair enough

**Comment by hjoaquim** [comment]
> > Thelp dialog for country will have all the choices regardless the provider - the choices should not be validated at the CLI (by argparse) but instead at the Platform (by pydantic); which seems to be happening as the image above shows.
> 
> This is actually not what happens with CLI, otherwise you would be able to enter: `economy --country us --provider econdb`.

Kind of; what happens is that the CLI merges the available choices on all the providers into a unique value list. Since `econdb` has no choices, the `oecd` ones prevail.
Not a showstopper imo, but definitively buggy - thanks for clarifying.

Very much like #6573 (tagging to reference this also on the issue)

**Review by hjoaquim — APPROVED** [review]
I'm good with this, thanks @deeleeramone for the clarifications!
Maybe would be good if @montezdesousa run through it also, since he was the one adding the `multiple_items_allowed` feature

File: openbb_platform/core/openbb_core/provider/standard_models/balance_sheet.py [hunk 8]
```diff
@@ -16,10 +16,6 @@ class BalanceSheetQueryParams(QueryParams):
     """Balance Sheet Query."""
 
     symbol: str = Field(description=QUERY_DESCRIPTIONS.get("symbol", ""))
-    period: str = Field(
-        default="annual",
-        description=QUERY_DESCRIPTIONS.get("period", ""),
-    )
```
Comment (montezdesousa) [inline openbb_platform/core/openbb_core/provider/standard_models/balance_sheet.py:22]: why are we removing `period` from this model?

File: openbb_platform/openbb/package/crypto_price.py [hunk 9]
```diff
@@ -135,11 +135,39 @@ def historical(
...
+                                "90m",
+                                "1h",
+                                "1d",
+                                "5d",
+                                "1W",
+                                "1M",
+                                "1Q",
+                            ],
+                        },
+                    },
```
Comment (montezdesousa) [inline openbb_platform/openbb/package/crypto_price.py:170]: Does this change in the `choices` affect the CLI autocomplete?

Comment (hjoaquim) [reply openbb_platform/openbb/package/crypto_price.py:170]: iiuc, no; although it won't pick from the `Field` anymore, it can still pick the choices from the reference.

File: openbb_platform/core/openbb_core/app/provider_interface.py (hunk 1 above)
Comment (deeleeramone) [reply openbb_platform/core/openbb_core/app/provider_interface.py:256]: Yes, this is a dictionary of choices for the field, by provider, where "multiple_items_allowed" is part of "choices".

**Comment by deeleeramone** [comment]
> Kind of; what happens is that the CLI merges the available choices on all the providers into a unique value list. Since `econdb` has no choices, the `oecd` ones prevail. Not a showstopper imo, but definitively buggy - thanks for clarifying.
> 
> Very much like #6573 (tagging to reference this also on the issue)

In *this specific instance* it is not entirely a show-stopper but it certainly has the potential to be. `derivatives.futures.curve` comes to mind.

No provider should "prevail" over another, all queries should make it to the provider for it to handle/validate specifically.

File: openbb_platform/core/openbb_core/provider/standard_models/balance_sheet.py (hunk 8 above)
Comment (deeleeramone) [reply openbb_platform/core/openbb_core/provider/standard_models/balance_sheet.py:22]: Because this is part of the issue, docstrings and type hints do not generate correctly. Long-standing issue being resolved.

**Review by IgorWounds — APPROVED** [review]

File: openbb_platform/core/openbb_core/app/provider_interface.py (hunk 1 above)
Comment (montezdesousa) [reply openbb_platform/core/openbb_core/app/provider_interface.py:256]: I also find this variable name ambiguous here because `choices` is also used with a different meaning as a key:value pair inside the dictionary

**Review by montezdesousa — APPROVED** [review]
